system/lib/fetch/emscripten_fetch.cpp

```diff
@@ -66,16 +68,58 @@ emscripten_fetch_t *emscripten_fetch(emscripten_fetch_attr_t *fetch_attr, const
 	}
 
 	emscripten_fetch_t *fetch = (emscripten_fetch_t *)malloc(sizeof(emscripten_fetch_t));
+    if ( !fetch ) return 0;
 	memset(fetch, 0, sizeof(emscripten_fetch_t));
 	fetch->id = globalFetchIdCounter++; // TODO: make this thread-safe!
 	fetch->userData = fetch_attr->userData;
-	fetch->url = strdup(url); // TODO: free
+    fetch->url = strdup( url );
+    if ( !fetch->url )
+    {
+        emscripten_fetch_free( fetch );
+        return 0;
+    }
 	fetch->__attributes = *fetch_attr;
-	fetch->__attributes.destinationPath = fetch->__attributes.destinationPath ? strdup(fetch->__attributes.destinationPath) : 0; // TODO: free
-	fetch->__attributes.userName = fetch->__attributes.userName ? strdup(fetch->__attributes.userName) : 0; // TODO: free
-	fetch->__attributes.password = fetch->__attributes.password ? strdup(fetch->__attributes.password) : 0; // TODO: free
-	fetch->__attributes.requestHeaders = 0;// TODO:strdup(fetch->__attributes.requestHeaders);
-	fetch->__attributes.overriddenMimeType = fetch->__attributes.overriddenMimeType ? strdup(fetch->__attributes.overriddenMimeType) : 0; // TODO: free
+
+#define SAFE_STRDUP(s)                  \
+  if (s)                                \
+  {                                     \
+    s = strdup(s);                      \
+    if (!s)                             \
+    {                                   \
+      emscripten_fetch_free(fetch);     \
+      return 0;                         \
+    }                                   \
+  }
+
+    SAFE_STRDUP( fetch->__attributes.destinationPath );
+    SAFE_STRDUP( fetch->__attributes.userName );
+    SAFE_STRDUP( fetch->__attributes.password );
+    if ( fetch->__attributes.requestHeaders )
+    {
+        size_t headersCount;
+        for ( headersCount = 0; fetch->__attributes.requestHeaders[ headersCount ]; ++headersCount );
+        const char** headers = ( const char** ) malloc( ( headersCount + 1 ) * sizeof( const char* ) );
+        if ( !headers )
+        {
+            emscripten_fetch_free( fetch );
+            return 0;
+        }
+        for ( size_t i = 0; i < headersCount; ++i )
+        {
+            headers[ i ] = strdup( fetch->__attributes.requestHeaders[ i ] );
+            if ( !headers[ i ] )
+            {
+                emscripten_fetch_free( fetch );
+                return 0;
+            }
+        }
+        headers[ headersCount ] = 0;
+        fetch->__attributes.requestHeaders = headers;
```

Comment:
Hmm, given that we only assign `fetch->__attributes.requestHeaders` here, the function `emscripten_fetch_free()` does not see any of these `strdup()`ed headers or the `malloc()`ed headers array above, and if any of these allocs fail, this will result in a memory leak.

To do safe cleanup, how about removing the `headers` variable altogether, and immediately assign the `malloc()` on top to `fetch->__attributes.requestHeaders`, grabbing the old pointer to a safe location first, and then immediately after `malloc()`, `memset()` the newly allocated block to zero, and then iterate to `strdup` all the contents? That way `emscripten_fetch_free()` will properly deallocate any partially allocated data on failure.

Reply:
I'm not sure if that's possible, `fetch->__attributes.requestHeaders` is a `const * char * const` so we can't directly assign to the individual const elements of the array:

```
const char * const *headers = fetch->__attributes.requestHeaders;
fetch->__attributes.requestHeaders = (const char * const*)malloc((headersCount + 1) * sizeof(const char*));
// ...null checking & memset
for(size_t i = 0; i < headersCount; ++i)
{
    fetch->__attributes.requestHeaders[i] = strdup(headers[i]); // <--- Can't do this
}
```

Instead I can ensure the `headers` array is freed up in the event of a failed alloc ?